### PR TITLE
API Token Limiting

### PR DESCRIPTION
**Pull request by farzadab**
This PR adds automatic handling of large API responses by chunking them and exposing a `loadBySimilarity` function to LLM based on Cohere reranker. The reranker is proxied through the Fixie server (PR https://github.com/fixie-ai/fixie/pull/1573).

There are still some TODOs that I will get rid of, but I would love to have reviewers' thoughts on it as well.

TODO:
- [x] remove and finalize TODOs
- [x] fix test errors
- [x] before merge: add changelog and update versions (`fixie-sdk` and `ai-jsx`)

image

### CODE REVIEW

File: packages/ai-jsx/src/batteries/sidekick/platform/large-response-handler.tsx
Comment (farzadab) [inline packages/ai-jsx/src/batteries/sidekick/platform/large-response-handler.tsx:55]: For some reason I didn't see any issues with rendering tool responses that might be an object (i.e. parsed JSON vs stringified).

@petersalas since this might be relevant to your existing PR, just wanted to get you to double check this part.

https://fixie-ai.slack.com/archives/C05B6EE9JE5/p1694191388802549

File: packages/ai-jsx/src/batteries/sidekick/platform/conversation.tsx
Comment (farzadab) [inline packages/ai-jsx/src/batteries/sidekick/platform/conversation.tsx:100]: This might be confusing, but here `maxLength` is based on characters, but in `LargeFunctionResponseHandler` it's based on tokens by default (i.e. `lengthFunction` is set to `tokenCounter`). I wanted to allow both chars and tokens here but the effort didn't seem worth it at the time. I can take another pass at it though if need be.

File: packages/ai-jsx/src/batteries/sidekick/platform/large-response-handler.tsx
Comment (farzadab) [inline packages/ai-jsx/src/batteries/sidekick/platform/large-response-handler.tsx:17]: I'm arbitrarily using OpenAI's tokenizer here. Trying to tie this back to the actual model being used didn't seem worth it.

File: packages/ai-jsx/src/batteries/sidekick/platform/large-response-handler.tsx
Comment (farzadab) [inline packages/ai-jsx/src/batteries/sidekick/platform/large-response-handler.tsx:45]: If we want to optimize for Cohere, we can use their tokenizer and set the `chunkLimit` to 500 (that way the capacity of reranker becomes 500K tokens (less the overlap).

File: packages/ai-jsx/src/batteries/sidekick/platform/large-response-handler.tsx
Comment (farzadab) [inline packages/ai-jsx/src/batteries/sidekick/platform/large-response-handler.tsx:94]: I'd rather not to do this, but @juberti if you still want it I can take another stab at it.

File: packages/ai-jsx/src/batteries/sidekick/platform/conversation.tsx
Comment (NickHeiner) [inline packages/ai-jsx/src/batteries/sidekick/platform/conversation.tsx:76]: minor
```suggestion
  return _.takeRightWhile(fullConversation, ({type}) => type !== 'user');
```

File: packages/ai-jsx/src/batteries/sidekick/platform/conversation.tsx
Comment (NickHeiner) [inline packages/ai-jsx/src/batteries/sidekick/platform/conversation.tsx:108]: Are the types wrong here?

File: packages/ai-jsx/src/batteries/sidekick/platform/large-response-handler.tsx
Comment (NickHeiner) [inline packages/ai-jsx/src/batteries/sidekick/platform/large-response-handler.tsx:11]:
```suggestion
export type LengthFunction = (text: string) => Promisable<number>
```

https://github.com/sindresorhus/type-fest/blob/main/source/promisable.d.ts

File: packages/ai-jsx/src/batteries/sidekick/platform/large-response-handler.tsx
Comment (NickHeiner) [inline packages/ai-jsx/src/batteries/sidekick/platform/large-response-handler.tsx:77]:
```suggestion
  // Option 2: try dumping as YAML. If it's small enough, then we are done.
```

uber-nit

File: packages/ai-jsx/src/batteries/sidekick/platform/large-response-handler.tsx
Comment (NickHeiner) [inline packages/ai-jsx/src/batteries/sidekick/platform/large-response-handler.tsx:83]: Hmm. I feel like, if the Cohere API key isn't set, it would be better to just truncate the response.

File: packages/ai-jsx/src/batteries/sidekick/platform/large-response-handler.tsx
Comment (NickHeiner) [inline packages/ai-jsx/src/batteries/sidekick/platform/large-response-handler.tsx:166]:
```suggestion
  return lastFnResData;
```

Maybe I'm missing something, but that first condition seems unnecessary.

File: packages/ai-jsx/src/batteries/sidekick/platform/large-response-handler.tsx
Comment (NickHeiner) [inline packages/ai-jsx/src/batteries/sidekick/platform/large-response-handler.tsx:160]: Hmm, it's a little odd that the function is named `lastFnResData`, but this loop will continue searching backwards until it finds the first element meeting the criteria. Did you want a `findLast` call instead?

File: packages/ai-jsx/src/batteries/sidekick/platform/large-response-handler.tsx
Comment (NickHeiner) [inline packages/ai-jsx/src/batteries/sidekick/platform/large-response-handler.tsx:169]: Do we ever anticipate having more than one tool? Having machinery for multiple tools seems like overkill if we only ever actually use a single tool?

File: packages/ai-jsx/src/batteries/sidekick/platform/large-response-handler.tsx
Comment (NickHeiner) [inline packages/ai-jsx/src/batteries/sidekick/platform/large-response-handler.tsx:217]:
```suggestion
  let content: Jsonifiable;
```

from `type-fest`

File: packages/ai-jsx/src/lib/cohere.tsx
Comment (NickHeiner) [inline packages/ai-jsx/src/lib/cohere.tsx:62]:
```suggestion
    // fall back to original order
```

"fallback" is a noun; "fall back" is a verb

File: packages/ai-jsx/src/batteries/sidekick/platform/large-response-handler.tsx
Comment (farzadab) [reply packages/ai-jsx/src/batteries/sidekick/platform/large-response-handler.tsx:169]: We used to have `loadByExpression` and `loadRemaining` (e.g. the functions can be conditioned based on the nature of the results), so the idea was to allow adding that back in if needed.

Also, handling 0 vs 1 tool seemed easier this way which also allows for more tools later on.

File: packages/ai-jsx/src/lib/cohere.tsx
Comment (NickHeiner) [inline packages/ai-jsx/src/lib/cohere.tsx:64]:
```suggestion
    const countResultDocs = Math.min(documents.length, top_n);
    return _.range(countResultDocs).map(index => ({ index, relevance_score: 0 }))
```

File: packages/ai-jsx/src/batteries/sidekick/platform/conversation.tsx
Comment (farzadab) [reply packages/ai-jsx/src/batteries/sidekick/platform/conversation.tsx:108]: I'm not sure why TypeScript doesn't throw an error here. There is no reason why `name` should be available in `tools`.
I also noticed that @petersalas had handled the case where `func` was `undefined`, but from what I understand, this line will error out before getting to that point since `tool[name]` can be the thing that is `undefined` (usually if the tool exists, the `func` also exists, but no guarantee that the `tool` exists).

File: packages/ai-jsx/src/batteries/sidekick/platform/large-response-handler.tsx
Comment (farzadab) [reply packages/ai-jsx/src/batteries/sidekick/platform/large-response-handler.tsx:83]: > I feel like, if the Cohere API key isn't set

We do set it using a context inside `request-wrapper.tsx`, so Fixie sidekicks will have it set. Also we do have a default context that uses `COHERE_API_KEY` env var. Is there anywhere I should document this?

File: packages/ai-jsx/src/batteries/sidekick/platform/large-response-handler.tsx
Comment (farzadab) [reply packages/ai-jsx/src/batteries/sidekick/platform/large-response-handler.tsx:166]: Huh you're right.

File: packages/ai-jsx/src/batteries/sidekick/platform/large-response-handler.tsx
Comment (farzadab) [reply packages/ai-jsx/src/batteries/sidekick/platform/large-response-handler.tsx:160]: Yes. At first this was implemented with a `.findLastIndex`, but type checking issues and I didn't want to have duplicate code. I'll take another look since I see I'm already doing `as unknown as RedactedFuncionResponseMetadata` anyway.

File: packages/ai-jsx/src/batteries/sidekick/platform/large-response-handler.tsx
Comment (farzadab) [reply packages/ai-jsx/src/batteries/sidekick/platform/large-response-handler.tsx:11]: But LC uses this signature and TypeScript doesn't understand that they are the same 😢

File: packages/ai-jsx/src/batteries/sidekick/platform/large-response-handler.tsx
Comment (farzadab) [reply packages/ai-jsx/src/batteries/sidekick/platform/large-response-handler.tsx:83]: cc @hessamb

Comment (hessamb) [reply packages/ai-jsx/src/batteries/sidekick/platform/large-response-handler.tsx:85]: I think that's a good idea. But falling back to truncate by char is also a good interim solution for hackathon (since we expect cohere API to be set anyway).

File: packages/ai-jsx/src/batteries/sidekick/platform/large-response-handler.tsx
Comment (NickHeiner) [reply packages/ai-jsx/src/batteries/sidekick/platform/large-response-handler.tsx:12]: Bummer. Ok, that's fine.

What if we tried this then?

```suggestion
export type LengthFunction = ((text: string) => number | Promise<number>);
```

**Comment by NickHeiner** [comment]
No need to block this PR on my approval.

File: packages/ai-jsx/src/batteries/sidekick/platform/large-response-handler.tsx
Comment (farzadab) [reply packages/ai-jsx/src/batteries/sidekick/platform/large-response-handler.tsx:12]: Same issue.

I might end up ejecting (i.e. copy&pasting) the text splitter from LC. If I do that I'll make sure to use `Promisable` instead.

File: packages/ai-jsx/src/batteries/sidekick/platform/large-response-handler.tsx
Comment (farzadab) [reply packages/ai-jsx/src/batteries/sidekick/platform/large-response-handler.tsx:53]: No real reason. I just thought that if there is an error, we might not want to pollute the context window with it. This might not be an important concern since the effects won't be linear due to `Shrinkable`.

File: packages/ai-jsx/src/lib/cohere.tsx
Comment (petersalas) [inline packages/ai-jsx/src/lib/cohere.tsx:66]: This seems like the wrong place for this fallback -- it seems like this function should probably just be the wrapper around the API?

File: packages/ai-jsx/src/core/conversation.tsx
Comment (petersalas) [inline packages/ai-jsx/src/core/conversation.tsx:114]: Assuming it's not a large number of places that needs this, did you consider `AI.PropsOfComponent<typeof FunctionResponse>`?

File: packages/ai-jsx/src/batteries/sidekick/platform/large-response-handler.tsx
Comment (petersalas) [inline packages/ai-jsx/src/batteries/sidekick/platform/large-response-handler.tsx:137]: I think using `renderToConversation` and then replacing just the `<FunctionResponse>`s  is preferable. Would rather minimize the proliferation of partial rendering.

File: packages/ai-jsx/src/batteries/sidekick/platform/large-response-handler.tsx
Comment (petersalas) [inline packages/ai-jsx/src/batteries/sidekick/platform/large-response-handler.tsx:105]: Will having multiple such `<FunctionResponse>s` in the conversation be a problem?

File: packages/ai-jsx/src/batteries/sidekick/platform/large-response-handler.tsx
Comment (petersalas) [reply packages/ai-jsx/src/batteries/sidekick/platform/large-response-handler.tsx:19]: I guess the dumbest/simplest option to consider would just be to use string length?

File: packages/fixie-sdk/src/request-wrapper.tsx
Comment (petersalas) [inline packages/fixie-sdk/src/request-wrapper.tsx:6]: will need to rev fixie-sdk as well

File: packages/ai-jsx/src/batteries/sidekick/platform/large-response-handler.tsx
Comment (farzadab) [reply packages/ai-jsx/src/batteries/sidekick/platform/large-response-handler.tsx:105]: I don't think so. We only present the model with the last one, and LLM gets at least one turn to react to it (also the `loadBySimilarity` itself will not trigger redaction so it can be called repeatedly). Is there any other scenario you were thinking about?

File: packages/ai-jsx/src/batteries/sidekick/platform/large-response-handler.tsx
Comment (farzadab) [reply packages/ai-jsx/src/batteries/sidekick/platform/large-response-handler.tsx:19]: Hmm, I agree that that's more natural and less opinionated (which is preferred for a library). In practice though I kinda want to use OpenAI's tokenizer for sidekicks.

Specially because we probably want to maintain the parity between chunks from a corpus and chunks from an API (easier to reason about I guess?).


Idea: What if we push `tokenizer`, `lengthFunction`, and such to the context too?

File: packages/ai-jsx/src/lib/cohere.tsx
Comment (farzadab) [reply packages/ai-jsx/src/lib/cohere.tsx:66]: You're right. Removed the fallback altogether.

File: packages/ai-jsx/src/batteries/sidekick/platform/large-response-handler.tsx
Comment (petersalas) [reply packages/ai-jsx/src/batteries/sidekick/platform/large-response-handler.tsx:19]: To be clear I agree with you that just using OpenAI is the right choice vs. doing largely performative plumbing -- was just throwing out an "even simpler" option for you to (re)consider :)

File: packages/ai-jsx/src/batteries/sidekick/platform/large-response-handler.tsx
Comment (petersalas) [reply packages/ai-jsx/src/batteries/sidekick/platform/large-response-handler.tsx:105]: WDYM we only present the model with the last one -- won't this be persisted in the conversation? And re:side-by-side:

- user
- functionCall
- functionResponse
- functionCall
- functionResponse
- assistant

is certainly a valid scenario, but I'm likely misunderstanding something.

A reasonable answer might be just that the model is smart enough to infer that the last one takes precedence over previous ones, but I was curious if this is something that either can't happen (i.e. I'm misunderstanding things) or you've played around with and is fine.

File: packages/ai-jsx/src/batteries/sidekick/platform/large-response-handler.tsx
Comment (farzadab) [reply packages/ai-jsx/src/batteries/sidekick/platform/large-response-handler.tsx]: Done.

File: packages/ai-jsx/src/batteries/sidekick/platform/large-response-handler.tsx
Comment (farzadab) [reply packages/ai-jsx/src/batteries/sidekick/platform/large-response-handler.tsx:133]: Ooh sorry yeah you're right. I'm assuming that the model is smart enough. I have certainly used multiple of these in the same conversation and didn't have any issues.

But this is a valid point. I'll play with the descriptions a bit more to make it a bit more obvious.

Comment (farzadab) [reply packages/ai-jsx/src/batteries/sidekick/platform/large-response-handler.tsx:133]: I improved the function description. For now I don't have better ideas to make sure this doesn't happen. I say we can assume that it works until we see any issues.
I put a TODO in the code to track this concern.

WDYT?

File: packages/ai-jsx/src/batteries/docs.tsx
Comment (farzadab) [inline packages/ai-jsx/src/batteries/docs.tsx:758]: @mdepinet here is my workaround fix for the type issue with newer versions of LC. The issue is that `VectorStore` doesn't take a single `filter` type as input anymore. It's dependent on the type of store that you use, e.g. Pinecone still takes a dictionary but `MemoryStore` takes a filter function as input and it throws an error if we provide an `object`.

Unfortunately, because of the way they do the typing, we don't get any TypeScript errors either way (so it can't easily be handled at compile time) and at runtime the types don't exist so we cannot decide what to pass in.
I decided to kinda put the burden on the user by saying that `params.filter` is what they provide and if it doesn't work then it's their fault.
This makes the interface a bit weird, but since nobody is currently using it, this should probably be fine. We can revisit later. WDYT?

Comment (mdepinet) [reply packages/ai-jsx/src/batteries/docs.tsx:758]: Seems like a reasonable approach to me. At the end of the day, we can only do so much to spare people from LC's garbage types if they want to use LC.

File: packages/fixie-sdk/src/request-wrapper.tsx
Comment (farzadab) [reply packages/fixie-sdk/src/request-wrapper.tsx:6]: Done. Updated package versions (`ai-jsx` and `fixie-sdk`) and Changelog.

**Comment by farzadab** [comment]
PTAL, if there are no more comments, I'd like to merge this.

File: packages/ai-jsx/src/lib/cohere.tsx
Comment (petersalas) [inline packages/ai-jsx/src/lib/cohere.tsx:80]: nit: should be `logger.debug` or `logger.trace`

File: packages/ai-jsx/src/batteries/sidekick/platform/large-response-handler.tsx
Comment (petersalas) [reply packages/ai-jsx/src/batteries/sidekick/platform/large-response-handler.tsx:133]: Works for me

**Review by petersalas — APPROVED** [review]

File: packages/ai-jsx/src/batteries/sidekick/platform/large-response-handler.tsx
Comment (farzadab) [reply packages/ai-jsx/src/batteries/sidekick/platform/large-response-handler.tsx:74]: I'm resolving this to merge the PR, but if you have concerns let's address them in follow up PRs.

**Pull request merged**
